### PR TITLE
net.box: fix IO error messages

### PR DESCRIPTION
**Pull request by locker**
Commit c13b3a31a9b04d4e3fbf0be88f76285b2cc103a7 changed error messages reported by net.box on IO errors, which caused integration/cartridge test failure. This PR restores the old error messages: `strerror(error->saved_errno)` if `saved_errno` is set, `error->errmsg` otherwise.

### CODE REVIEW

File: src/lib/core/exception.cc [hunk 1]
```diff
@@ -172,6 +174,15 @@ SocketError::SocketError(const char *file, unsigned line,
 	errno = save_errno;
 }
 
+const struct type_info type_GaiError =
+	make_type("GaiError", &type_GaiError);
```
Comment (Gerold103) [inline src/lib/core/exception.cc:178]: You didn't add `GaiError` support to `mp_error`. Its type `"GaiError"` is lost when marshalled through network. Maybe you could reuse `SystemError` here. Perhaps inline what you did in the constructor? Also could keep errno as `EIO` and change only the error message. Errno does not seem to be relevant to the problem.

File: src/lib/core/exception.cc [hunk 2]
```diff
@@ -172,6 +174,15 @@ SocketError::SocketError(const char *file, unsigned line,
...
+
+GaiError::GaiError(const char *file, unsigned line, int errcode)
+	: SystemError(&type_SocketError, file, line)
+{
+	saved_errno = 0;
+	error_format_msg(this, "getaddrinfo: %s", gai_strerror(errcode));
```
Comment (Gerold103) [inline src/lib/core/exception.cc:184]: You didn't add `GaiError` support to `mp_error`. Its type `"GaiError"` is lost when marshalled through network. Maybe you could reuse `SystemError` here. Perhaps inline what you did in the constructor?

I don't really understand why do you need a new error type at all. Why did it work fine in `net_box.lua` before? Can we just do the same what was done there?

Comment (locker) [reply src/lib/core/exception.cc:184]: Thanks for tipping me about `mp_error` - I totally forgot about it. Fixed - see the diff below. We don't need any special treatment of `GaiError`, because all it does is set an error message, which is restored for all kinds of errors.

<details>
<summary>diff</summary>

```diff
diff --git a/src/box/mp_error.cc b/src/box/mp_error.cc
index fba562a84afd..7a3438511c1b 100644
--- a/src/box/mp_error.cc
+++ b/src/box/mp_error.cc
@@ -257,6 +257,8 @@ error_build_xc(struct mp_error *mp_error)
 		err = new SwimError();
 	} else if (strcmp(mp_error->type, "CryptoError") == 0) {
 		err = new CryptoError();
+	} else if (strcmp(mp_error->type, "GaiError") == 0) {
+		err = new GaiError();
 	} else {
 		err = new ClientError();
 	}
diff --git a/test/unit/mp_error.cc b/test/unit/mp_error.cc
index 5e68b34d29f9..42902f76c3d7 100644
--- a/test/unit/mp_error.cc
+++ b/test/unit/mp_error.cc
@@ -86,6 +86,7 @@ const char *standard_errors[] = {
 	"CollationError",
 	"SwimError",
 	"CryptoError",
+	"GaiError",
 };
 
 enum {
diff --git a/test/unit/mp_error.result b/test/unit/mp_error.result
index 356d2dc970ad..b4540252e2cf 100644
--- a/test/unit/mp_error.result
+++ b/test/unit/mp_error.result
@@ -1,7 +1,7 @@
 	*** main ***
 1..7
 	*** test_stack_error_decode ***
-    1..17
+    1..18
     ok 1 - check CustomError
     ok 2 - check AccessDeniedError
     ok 3 - check ClientError
@@ -18,7 +18,8 @@
     ok 14 - check CollationError
     ok 15 - check SwimError
     ok 16 - check CryptoError
-    ok 17 - stack size
+    ok 17 - check GaiError
+    ok 18 - stack size
 ok 1 - subtests
 	*** test_stack_error_decode: done ***
 	*** test_decode_unknown_type ***
```

</details>

> I don't really understand why do you need a new error type at all. Why did it work fine in net_box.lua before? Can we just do the same what was done there?

It worked before, because Lua's `socket.tcp_connect` doesn't use diag for error propagation. Instead, it sets an error message to `gai_strerror(code)` for gai errors and to `strerror(errno)` for all other kinds of errors. `coio_connect` does set diag, which is fine, because it's a C function. The problem is how a `SystemError` is set: `errmsg` usually explains what failed (e.g. `connect` or `recv`, see `sio.c`) while the reason of the failure is encoded in `saved_errno`. In the new version I simply returned `errmsg`, which resulted in confusing error messages. We should use `strerror(saved_errno)` instead. However, gai errors are different - they set `saved_errno` to `EIO`, while the actual reason of failure is written to `errmsg` so for them we need to use `errmsg`. To distinguish gai errors from other system errors, I added `GaiError` class, which, in contrast to `SystemError`, sets `saved_errno` to 0 so that in net.box code I can return `strerror(saved_errno)` if `saved_errno` isn't 0, and `errmsg` otherwise. This will also work when we introduce `TlsError` - the latter won't set `errno` so handling of `iostream_read` errors done in this patch will work for both plain streams, which set `SocketError`, and encrypted streams, which set `TlsError`.

This all looks pretty ugly, but I don't know how to fix it in a better way. We should probably add a virtual method for reporting the error reason and use it instead of `errmsg` everywhere. By default the method would return `errmsg`, but for system errors it would return `strerror(saved_errno)`. Anyway, we'll still need `GaiError` so I decided to do a quick fix without adding a virtual method for now. WDYT?

Comment (Gerold103) [reply src/lib/core/exception.cc:184]: I totally understand `TlsError`. But `GaiError` looks strange. I am not against it. I am more against adding new C++ classes, because I hope to drop C++ from the errors code entirely.

Instead of error reason separated we could just encode it into the error message. In addition to what failed also say why it failed, in the same string. However irrelevant here maybe.

To your patch - did you check it fixed the integration? I tried what happens in 1.10 (old netbox), master, and this PR. In all cases I did this:
```Lua
netbox = require('net.box')
c = netbox.connect('invalid host', 123)
```
**Master**:
```
tarantool> c
---
- host: invalid host
  state: error
  error: 'getaddrinfo: nodename nor servname provided, or not known'
  port: 123
...
```
**1.10**:
```
tarantool> c
---
- host: invalid host
  state: error
  error: Invalid argument
  port: 123
...
```
**Your PR**:
```
tarantool> c
---
- host: invalid host
  state: error
  error: 'getaddrinfo: nodename nor servname provided, or not known'
  port: 123
...
```
What changed? I thought the goal was to make it again as in 1.10.

Comment (locker) [reply src/lib/core/exception.cc:184]: > Instead of error reason separated we could just encode it into the error message. In addition to what failed also say why it failed, in the same string.

This would result in an error message like this:

```
connect, called on fd 1165, aka 127.0.0.1:52424: Connection refused
```

I think it's fine, but cartridge tests don't expect it:

```
...ad/src/tarantool/cartridge/test/integration/ddl_test.lua:270: Error message does not contain: "\"localhost:13301\": Connection refused"
Error message received: "...ad/src/tarantool/cartridge/test/integration/ddl_test.lua:37: \"localhost:13301\": connect, called on fd 50, aka 127.0.0.1:60448: Connection refused"
```

We could probably patch the tests though. We never promised a particular error message format AFAIU. Do you think it'd be fine?

> To your patch - did you check it fixed the integration?

Yes, it does fix integration - I checked manually. (Soon we'll have integration tests run automatically if the `full-ci` label is set - https://github.com/tarantool/tarantool/pull/6645.)

> I tried what happens in 1.10 (old netbox), master, and this PR.
> ...
> What changed? I thought the goal was to make it again as in 1.10. Or am I looking at a wrong test?

This is the only error message that differs and it changed long before my patches, at least in 2.8 so it should be okay.

Comment (locker) [reply src/lib/core/exception.cc:184]: I did what you suggested, see https://github.com/tarantool/tarantool/pull/6653, and frankly I like it more. This wouldn't fix integration/cartridge tests, but this would make error messages more useful. We can fix integration/cartridge tests by making them do pattern match instead of string comparison. Thanks for the suggestion and PTAL.

Comment (Gerold103) [reply src/lib/core/exception.cc:184]: > We could probably patch the tests though. We never promised a particular error message format AFAIU. Do you think it'd be fine?

It think it is good. Cartridge tests should not depend on error messages being incomplete. And we definitely should not avoid improvements just because of that. I had the same problem in vshard and managed to fix the tests just fine. Cartridge guys should do the same. Either match error code or match some parts of the error message.

IMO we only need to keep error codes when it is not too hard and error types (`"ClientError"`, `"SocketError"`, etc) for important errors like `box.error.READONLY`. They are not documented too AFAIK, but people could have some logic depending on them. Anyway, irrelevant here.
